File: capi/controller/internal/resources/workers.go
package resources

import (
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime/schema"
)

var (
	kubeadmTemplateGVK   = schema.GroupVersionKind{Group: "bootstrap.cluster.x-k8s.io", Version: "v1beta2", Kind: "KubeadmConfigTemplate"}
	devMachineGVK        = schema.GroupVersionKind{Group: "infrastructure.cluster.x-k8s.io", Version: "v1beta2", Kind: "DevMachineTemplate"}
	machineDeploymentGVK = schema.GroupVersionKind{Group: "cluster.x-k8s.io", Version: "v1beta2", Kind: "MachineDeployment"}
)

func KubeadmConfigTemplate(context Context) *unstructured.Unstructured {
	spec := map[string]any{
		"template": map[string]any{
			"spec": map[string]any{
				"joinConfiguration": map[string]any{
					"nodeRegistration": map[string]any{
						"kubeletExtraArgs": []any{
							map[string]any{
								"name":  "eviction-hard",
								"value": "nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%",
							},
						},
					},
				},
			},
		},
	}
	if len(context.WorkerBootstrapCommands) != 0 {
		commands := make([]any, 0, len(context.WorkerBootstrapCommands))
		for _, command := range context.WorkerBootstrapCommands {
			commands = append(commands, command)
		}
		spec["template"].(map[string]any)["spec"].(map[string]any)["preKubeadmCommands"] = commands
	}
	return object(context, kubeadmTemplateGVK, context.Tenant.Name, context.Tenant.Name+"-worker", "kubeadm-config-template", spec)
}

func DevMachineTemplate(context Context) *unstructured.Unstructured {
	return object(context, devMachineGVK, context.Tenant.Name, context.Tenant.Name+"-worker", "dev-machine-template", map[string]any{
		"template": map[string]any{
			"spec": map[string]any{
				"backend": map[string]any{
					"docker": map[string]any{
						"customImage":      context.Inputs.NodeImage,
						"bootstrapTimeout": "5m",
						"extraMounts": []any{
							map[string]any{
								"hostPath":      context.Inputs.CacheHostPath,
								"containerPath": context.Inputs.CacheContainerPath,
								"readOnly":      true,
							},
							map[string]any{
								"hostPath":      context.VolumePath,
								"containerPath": context.Inputs.StorageContainerPath,
								"readOnly":      false,
							},
						},
					},
				},
			},
		},
	})
}

func MachineDeployment(context Context) *unstructured.Unstructured {
	labels, annotations := markers(context, "machine")
	templateLabels := stringMap(labels)
	templateLabels["cluster.x-k8s.io/cluster-name"] = context.Tenant.Name
	templateLabels["cnpg-vcluster.capi/nodepool"] = "worker"
	return object(context, machineDeploymentGVK, context.Tenant.Name, context.Tenant.Name+"-worker", "machine-deployment", map[string]any{
		"clusterName": context.Tenant.Name,
		"replicas":    int64(context.Spec.Workers),
		"machineNaming": map[string]any{
			"template": "{{ .cluster.name }}-worker-{{ .random }}",
		},
		"selector": map[string]any{
			"matchLabels": map[string]any{
				"cluster.x-k8s.io/cluster-name": context.Tenant.Name,
				"cnpg-vcluster.capi/nodepool":   "worker",
			},
		},
		"template": map[string]any{
			"metadata": map[string]any{
				"labels":      templateLabels,
				"annotations": stringMap(annotations),
			},
			"spec": map[string]any{
				"clusterName": context.Tenant.Name,
				"version":     "v" + context.Spec.KubernetesVersion,
				"bootstrap": map[string]any{
					"configRef": map[string]any{
						"apiGroup": "bootstrap.cluster.x-k8s.io",
						"kind":     "KubeadmConfigTemplate",
						"name":     context.Tenant.Name + "-worker",
					},
				},
				"infrastructureRef": map[string]any{
					"apiGroup": "infrastructure.cluster.x-k8s.io",
					"kind":     "DevMachineTemplate",
					"name":     context.Tenant.Name + "-worker",
				},
			},
		},
	})
}

File: capi/controller/src/resources/workers.rs
use kube::core::DynamicObject;
use serde_json::json;
use std::collections::BTreeMap;

use super::Context;
use crate::ownership::{
    CLUSTER_API_VERSION, FOUNDATION_ANNOTATION, SPEC_HASH_ANNOTATION, TENANT_UID_ANNOTATION,
};

pub fn kubeadm_config_template(context: &Context<'_>) -> DynamicObject {
    let mut spec = json!({"template":{"spec":{"joinConfiguration":{"nodeRegistration":{"kubeletExtraArgs":[{
        "name":"eviction-hard","value":"nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%"
    }]}}}}});
    if !context.worker_bootstrap_commands.is_empty() {
        spec["template"]["spec"]["preKubeadmCommands"] = json!(context.worker_bootstrap_commands);
    }
    context.object(
        "bootstrap.cluster.x-k8s.io/v1beta2",
        "KubeadmConfigTemplate",
        &format!("{}-worker", context.name()),
        context.name(),
        "kubeadm-config-template",
        spec,
    )
}

pub fn dev_machine_template(context: &Context<'_>) -> DynamicObject {
    context.object("infrastructure.cluster.x-k8s.io/v1beta2", "DevMachineTemplate",
        &format!("{}-worker",context.name()), context.name(), "dev-machine-template", json!({
            "template":{"spec":{"backend":{"docker":{
                "customImage":context.inputs.node_image,"bootstrapTimeout":"5m",
                "extraMounts":[
                    {"hostPath":context.inputs.cache_host_path,"containerPath":context.inputs.cache_container_path,"readOnly":true},
                    {"hostPath":context.volume_path,"containerPath":context.inputs.storage_container_path,"readOnly":false}
                ]
            }}}}
        }))
}

pub fn machine_deployment(context: &Context<'_>) -> DynamicObject {
    let name = format!("{}-worker", context.name());
    let mut labels = context.identity().labels();
    labels.insert(
        "cluster.x-k8s.io/cluster-name".into(),
        context.name().into(),
    );
    labels.insert("cnpg-vcluster.capi/nodepool".into(), "worker".into());
    context.object(CLUSTER_API_VERSION, "MachineDeployment", &name, context.name(), "machine-deployment", json!({
        "clusterName":context.name(),"replicas":context.spec.workers,
        "machineNaming":{"template":"{{ .cluster.name }}-worker-{{ .random }}"},
        "selector":{"matchLabels":{"cluster.x-k8s.io/cluster-name":context.name(),"cnpg-vcluster.capi/nodepool":"worker"}},
        "template":{
            "metadata":{"labels":labels,"annotations":context.identity().annotations("machine")},
            "spec":{
                "clusterName":context.name(),"version":format!("v{}",context.spec.kubernetes_version),
                "bootstrap":{"configRef":{"apiGroup":"bootstrap.cluster.x-k8s.io","kind":"KubeadmConfigTemplate","name":name}},
                "infrastructureRef":{"apiGroup":"infrastructure.cluster.x-k8s.io","kind":"DevMachineTemplate","name":name}
            }
        }
    }))
}

pub fn storage_volume_name(context: &Context<'_>) -> String {
    format!("{}-{}-storage", context.inputs.lab_prefix, context.name())
}

pub fn storage_volume_labels(context: &Context<'_>) -> BTreeMap<String, String> {
    let mut labels = context.identity().labels();
    for (key, value) in [
        ("cnpg-vcluster.capi/role", "tenant-storage"),
        ("cnpg-vcluster.capi/tenant", context.name()),
        (TENANT_UID_ANNOTATION, context.identity().tenant_uid),
        (SPEC_HASH_ANNOTATION, context.spec_hash),
        (FOUNDATION_ANNOTATION, context.foundation_hash),
    ] {
        labels.insert(key.into(), value.into());
    }
    labels
}
